src/Core/ScreenshotCache.vala

```diff
@@ -0,0 +1,273 @@
+            return -1;
+        }
+    }
+
+    // Delete the oldest files in the screenshot cache until the cache is less than the max size.
+    private void delete_oldest_files (int screenshot_usage) {
+        string? name = null;
+        Dir dir;
+        if (read_dir (out dir, screenshot_path) == -1) {
+            return;
+        }
+
+        while (screenshot_usage > MAX_CACHE_SIZE) {
+            string? oldest_path = null;
+            time_t oldest_time = 0;
+            while ((name = dir.read_name ()) != null) {
+                string entry = Path.build_filename (screenshot_path, name);
+                Stat fstat = Stat (entry);
+                if (oldest_time == 0 || fstat.st_mtime < oldest_time) {
+                    oldest_time = fstat.st_mtime;
+                    oldest_path = entry;
+                }
+            }
+
+            if (null != oldest_path) {
+                debug ("deleting screenshot at %s to free cache\n", oldest_path);
+
+                var file = File.new_for_path (oldest_path);
+                int size = query_size (oldest_path);
+                if (size == -1) {
+                    return;
+                }
+
+                if (delete_file (file, oldest_path) == 0) {
+                    screenshot_usage -= size;
+                } else {
+                    return;
+                }
+            }
+        }
+    }
+
+    // Get the combined size of the screenshot cache.
+    private int summarize_screenshot_usage () {
+        Dir dir;
+        if (read_dir (out dir, screenshot_path) == -1) {
+            return -1;
+        }
+        
+        string? name = null;
+        int size = 0;
+
+        while ((name = dir.read_name ()) != null) {
+            string entry = Path.build_filename (screenshot_path, name);
+            int file_size = query_size (entry);
+            if (file_size == -1) {
+                return -1;
+            }
+
+            size += file_size;
+        }
+
+        return size;
+    }
+
+    // Generate a screenshot path based on the URL to be fetched.
+    private string generate_screenshot_path (string url) {
+        int ext_pos = url.last_index_of (".");
+        string extension = url.slice ((long) ext_pos, (long) url.length);
+        if (extension.contains ("/")) {
+            extension = "";
+        }
+
+        return Path.build_filename (
+            screenshot_path,
+            Path.DIR_SEPARATOR_S,
+            "%02x".printf (url.hash ()) + extension
+        );
+    }
+
+    /*
+     * Fetches a screenshot in a background thread.
+     * 
+     * A result indicating the success (0) will be returned as the result upon completion.
+     */
+    public async int fetch (string url, out string out_file) {
+        SourceFunc callback = fetch.callback;
+        int result = 0;
+        string path = generate_screenshot_path (url);
+        var file = File.new_for_path (path);
+
+        new Thread<bool> ("fetching_screenshot", () => {
```

Comment:
Any specific reason to prefer a thread here instead of using all the `async` variants of the methods there?

Reply:
Sounds like that would complicate this implementation greatly, and increases the amount of code being run in the main thread that doesn't have to be.

Reply:
how so? it just means using the `yield` keyword pretty everywhere

Reply:
Though using `yield` would not have this process running purely in the background. Each callback would hog processing time from the main thread.